src/routes/router.ts

```diff
@@ -0,0 +1,111 @@
+import { IttyRouter } from 'itty-router';
+import { Middleware } from '../middleware/middleware';
+import { Context } from '../context';
+import { parseUrl } from '../utils/request';
+import responses from '../responses';
+import { Request as WorkerRequest } from './request';
+
+/**
+ * Simple wrapper around {@link IttyRouter} that allows us to do a middleware
+ *  approach with our routing.
+ * @see {Middleware}
+ */
+export class Router {
+  private itty = IttyRouter<WorkerRequest, [Context], Response>();
+
+  handle(request: Request, ctx: Context): Promise<Response> {
+    return this.itty.fetch(request, ctx);
+  }
+
+  options(endpoint: string, middlewares: Middleware[]): void {
+    const middlewareChain = buildMiddlewareChain(middlewares);
+
+    this.itty.options(endpoint, (req, ctx) => {
+      return callMiddlewareChain(middlewareChain, req, ctx);
+    });
+  }
+
+  head(endpoint: string, middlewares: Middleware[]): void {
+    const middlewareChain = buildMiddlewareChain(middlewares);
+
+    this.itty.head(endpoint, (req, ctx) => {
+      return callMiddlewareChain(middlewareChain, req, ctx);
+    });
+  }
+
+  get(endpoint: string, middlewares: Middleware[]): void {
+    const middlewareChain = buildMiddlewareChain(middlewares);
+
+    this.itty.get(endpoint, (req, ctx) => {
+      return callMiddlewareChain(middlewareChain, req, ctx);
+    });
+  }
+}
+
+type MiddlewareChain = (
+  request: WorkerRequest,
+  ctx: Context
+) => Promise<Response>;
+
+/**
+ * Builds a chain of middlewares to call. Chains them in the same order as they
+ *  are in the `middlewares` array.
+ */
+function buildMiddlewareChain(middlewares: Middleware[]): MiddlewareChain {
+  // root will be the very first middleware in the chain
+  let root: MiddlewareChain = () => {
+    throw new Error('reached the end of the middleware chain');
+  };
+
+  // Link the middlewares in reverse order for simplicity sakes
+  for (let i = middlewares.length - 1; i >= 0; i--) {
```

Comment:
I don't think a manual hardcoded for loop is needed.

Just go `for (let middleware of middlewares.reverse())`

Reply:
That method reverses the array in-place, I'm not sure if we wanna modify it in case we need to use it somewhere else in the future. I also don't think that'd be as performant as this

Reply:
It doesn't reverse the array in place, it returns a reversed array

Reply:
(the performance difference is negligent to be honest)